Lib/test/support/strace_helper.py

```diff
@@ -0,0 +1,173 @@
+            strace_returncode=-1,
+            python_returncode=-1,
+            _raw_events=f"error({reason},details={details}) = -1",
+            stdout=res.out if res else "",
+            stderr=res.err if res else "")
+
+    # Run strace, and get out the raw text
+    try:
+        res, cmd_line = run_python_until_end(
+            "-c",
+            textwrap.dedent(code),
+            __run_using_command=[_strace_binary] + strace_flags)
+    except OSError as err:
+        return _make_error("Caught OSError", err)
+
+    if check and res.rc:
+        res.fail(cmd_line)
+
+    # Get out program returncode
+    decoded = res.err.decode().strip()
+
+    output = decoded.rsplit("\n", 1)
+    if len(output) != 2:
+        return _make_error("Expected strace events and exit code line",
+                           decoded[-50:])
+
+    returncode_match = _returncode_regex.match(output[1])
+    if not returncode_match:
+        return _make_error("Expected to find returncode in last line.",
+                           output[1][:50])
+
+    python_returncode = int(returncode_match["returncode"])
+    if check and python_returncode:
+        res.fail(cmd_line)
+
+    return StraceResult(strace_returncode=res.rc,
+                        python_returncode=python_returncode,
+                        _raw_events=output[0],
+                        stdout=res.out,
+                        stderr=res.err)
+
+
+def _get_events(code: str, strace_flags: List[str], prelude: str,
+                cleanup: str) -> List[StraceEvent]:
+    # NOTE: The flush is currently required to prevent the prints from getting
+    # buffered and done all at once at exit
+    prelude = textwrap.dedent(prelude)
+    code = textwrap.dedent(code)
+    cleanup = textwrap.dedent(cleanup)
+    to_run = f"""
+print("MARK prelude", flush=True)
+{prelude}
+print("MARK code", flush=True)
+{code}
+print("MARK cleanup", flush=True)
+{cleanup}
+print("MARK __shutdown", flush=True)
+    """
+    trace = strace_python(to_run, strace_flags)
+    all_sections = trace.sections()
+    return all_sections['code']
+
+
+def get_syscalls(code: str,
+                 strace_flags: List[str],
+                 prelude: str = "",
+                 cleanup: str = "") -> List[str]:
+    """Get the syscalls which a given chunk of python code generates"""
+    events = _get_events(code, strace_flags, prelude=prelude, cleanup=cleanup)
+    return [ev.syscall for ev in events]
+
+
+def _can_strace():
+    res = strace_python("import sys; sys.exit(0)", [], check=False)
+    assert res.events(), "Should have parsed multiple calls"
+
+    global _strace_working
+    _strace_working = res.strace_returncode == 0 and res.python_returncode == 0
+
+
+def requires_strace():
+    if sys.platform != "linux":
+        return unittest.skip("Linux only, requires strace.")
+    # Moderately expensive (spawns a subprocess), so share results when possible.
+    if _strace_working is None:
+        _can_strace()
```

Comment:
move the assignment to the global _strace_working into this function so that the is None test and code assigning it to something other than that all lives together in one scope.  (have _can_strace() return a boolean to be assigned perhaps)